packages/frontend/src/features/metricsCatalog/hooks/useProjectTags.tsx

```diff
@@ -24,14 +26,78 @@ const createTag = async (
  */
 export const useCreateTag = () => {
     const queryClient = useQueryClient();
+
     return useMutation<
         ApiCreateTagResponse['results'],
         ApiError,
-        { projectUuid: string; data: Pick<Tag, 'name' | 'color'> }
+        {
+            projectUuid: string;
+            data: Pick<Tag, 'name' | 'color'>;
+            catalogSearchUuid?: string; // Make optional to support both with/without catalog item tagging
+        },
+        { previousTags: Tag[] }
     >({
-        mutationFn: ({ projectUuid, data }) => createTag(projectUuid, data),
+        // @ts-expect-error - TODO: fix this
+        mutationFn: async ({ projectUuid, data, catalogSearchUuid }) => {
+            // First create the tag
+            const newTag = await createTag(projectUuid, data);
+
+            // If catalogSearchUuid is provided, also tag the catalog item
+            if (catalogSearchUuid) {
+                await addCategoryToCatalogItem({
+                    projectUuid,
+                    catalogSearchUuid,
+                    tagUuid: newTag.tagUuid,
+                });
+            }
+
+            return newTag;
+        },
+        onMutate: async ({ projectUuid, data }) => {
+            // Cancel any outgoing refetches
+            await queryClient.cancelQueries(['project-tags', projectUuid]);
+
+            // Snapshot the previous value
+            const previousTags = queryClient.getQueryData<Tag[]>([
+                'project-tags',
+                projectUuid,
+            ]);
+
+            // Optimistically update the cache
+            queryClient.setQueryData(
+                ['project-tags', projectUuid],
+                (old: Tag[] = []) => {
+                    const optimisticTag: Tag = {
```

Comment:
move outside of the `setQueryData` 
and use the `optimisticTag` to update the `metrics-catalog` query data as well!